giskard/llm/talk/tools/metric.py

```diff
@@ -0,0 +1,117 @@
+from difflib import SequenceMatcher as SeqM
+
+from giskard.datasets.base import Dataset
+from giskard.llm.talk.config import AVAILABLE_METRICS, FUZZY_SIMILARITY_THRESHOLD, ToolDescription
+from giskard.llm.talk.tools.base import BaseTool, get_feature_json_type
+
+
+class MetricTool(BaseTool):
+    """Performance metric calculation Tool.
+
+    Attributes
+    ----------
+    default_name : str
+        The default name of the Tool. Can be re-defined with constructor.
+    default_description: str
+        The default description of the Tool's functioning. Can be re-defined with constructor.
+    """
+
+    default_name: str = "calculate_metric"
+    default_description: str = ToolDescription.CALCULATE_METRIC.value
+
+    @property
+    def specification(self) -> str:
+        """Return the Tool's specification in a JSON Schema format.
+
+        Returns
+        -------
+        str
+            The Tool's specification.
+        """
+        feature_json_type = get_feature_json_type(self._dataset)
+
+        return {
+            "type": "function",
+            "function": {
+                "name": self.name,
+                "description": self.description,
+                "parameters": {
+                    "type": "object",
+                    "properties": {
+                        "metric_type": {"type": "string", "enum": list(AVAILABLE_METRICS.keys())},
+                        "features_dict": {
+                            "type": "object",
+                            "properties": {feature: {"type": dtype} for feature, dtype in feature_json_type.items()},
+                        },
+                    },
+                    "required": ["metric_type", "features_dict"],
+                },
+            },
+        }
+
+    def _get_input_from_dataset(self, row_filter: dict[str, any]) -> Dataset:
+        """Get input from dataset.
+
+        Filter rows from the dataset, using the `row_filter`.
+
+        Parameters
+        ----------
+        row_filter : dict[str, any]
+            The dictionary with features and related values to filter the dataset.
+
+        Returns
+        -------
+        Dataset
+            The Giskard Dataset with filtered rows.
+        """
+        filtered_df = self._dataset.df
```

Comment:
@rabah-khalek 

1. The tools are initialised within the "talk" function, for which the `model` and `dataset` are mandatory parameters. This means, that we won't have a situation, that the 'Metric' tool will lack the model or the dataset while being called.
2. For example, the 'Scan' tool do not use the model or dataset objects. That is why I made them optional for the BaseTool object.

I envision workarounds:
a) Remove 'optional' typing, but put default 'None' values to the **init**
b) Remove 'optional' typing, do not put default 'None' and just pass `model` and `dataset` to the 'Scan' tool.
c) Create separate base class for the 'Scan' tool (do not like it, since we create new class for a small change).

WDYT?

Reply:
For the moment I went for the second option `Remove 'optional' typing, do not put default 'None' and just pass model and dataset to the 'Scan' tool, even if they won't be used.`

Reply:
I would do this:
```py
class BaseTool(ABC):
    def __init__(
        self,
        name: Optional[str] = None,
        description: Optional[str] = None,
    ):
        self._name = name if name is not None else self.default_name
        self._description = description if description is not None else self.default_description

class Foo(BaseTool):
  def __init__(
        self,
        model: Optional[BaseModel] = None,
        dataset: Optional[Dataset] = None):
        super().__init__()
        self._model = model
        self._dataset = dataset

class Bar(BaseTool):
  def __init__(
        self,
        scan_report: Optional[ScanReport] = None):
        super().__init__()
        self._scan_report = scan_report
```